### ISSUE TITLE
[Translate][4.2] transchoice is deprecated, new way of doing things is missing

### ISSUE THREAD

**Issue by mhujer** [symfony/symfony-docs#11002]
`transChoice` method and `transchoice` filter are deprecated since 4.2, but the documentation does not show the example of the new way of doing things.

The note about the deprecation was added [to documentation](https://symfony.com/doc/current/translation.html#pluralization) in https://github.com/symfony/symfony-docs/pull/10973, but the examples still show the old way (using `transChoice`).

The example in the blogpost (https://symfony.com/blog/new-in-symfony-4-2-intlmessageformatter) is rather complicated.

**Comment by arjenm** [comment]
Also, don't forget to clearly add the translation files need to be renamed to include "+intl-icu" in the filename. 

It seems the deprecation message is additionally unclear since it requires a lot more than simply replacing the transchoice-call with a trans-call.

**Comment by mhujer** [comment]
@arjenm Did you manage to get the new approach working? I gave up after a while :-(

**Comment by arjenm** [comment]
Yes we did :)

But it was indeed a lot of trial and error and using resources that turned out to be linked in that blog post (and were much more helpful) :(

In your templates this is roughly what had to change:

```twig
Before:
{{ 'some.number_of_items'|transchoice(numberOfItems, {'%count%': numberOfItems}) }}
After:
{{ 'some.number_of_items'|trans({'count': numberOfItems}) }}
```

And in the translations.

Rename the files from something like messages.en.yml to messages+intl-icu.en.yml
If you rename it though, no old style transchoice in this file will work since the new translator simply does not understand it (which is also not documented!).

And change the contents:
```yaml
# Before
some:
  number_of_items: "{0} No items found.|{1} One item found.|]1,Inf[ %count% items found."
# After
some:
  number_of_items: "{count, plural,
        =0 {No items found.}
        one {One item found.}
        other {# items found.}
      }"
# Or without newlines and/or nested inside some sentence:	  
some:
  number_of_items: "{count, plural, =0 {No items found.} one {One item found.} other {# items found.}}"
  nested_inside_text: "some text before {count, plural, =0 {No items found.} one {One item found.} other {# items found.}} and after""
```
Note the 'one' and 'other' are special keys that indicate certain groups of numbers depending on the locale. There can be different groups per locale (i.e. some have one, two, few, many, other).
And since only a few locales have a special 'zero' class, you can also explicitly specify 0 (with that =0).

For more allowed formats and more information, these resources turned out to be quite helpful:
https://messageformat.github.io/messageformat/page-guide
http://userguide.icu-project.org/formatparse/messages
http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html

In the end the new variant is actually more powerful. As the blog post shows, you can combine and nest a mixture of plurals and 'selects'. I.e. the male/female/other example from the blog post. We also converted a use case for that, to test how it worked: we have a listing of a mixture of different article types and wanted to have a dynamic variation in both the naming of that type and its pluralization. So now we do something like: 'listing.article_type'|trans({'articleType': article.getType(), 'count': list.length()}), where we earlier had an ugly hack with a concat like this:
'listing.article_type' ~ article.getType()|transchoice(list.length(), {'%count%': list.length()}),

Which actually turned out to be quite similar to that complex example from, the blog post :P

**Comment by mhujer** [comment]
@arjenm thanks for the detailed guide, I will try, if I can make it work  👍

**Comment by mhujer** [comment]
@arjenm Thank you again! I just removed all `transchoice` filters from our application and your guide really helped me in the process 👏

**Comment by mhujer** [comment]
Surprisingly, [translation placeholders with `trans` method](https://symfony.com/doc/current/components/translation/usage.html#component-translation-placeholders) stop working when the new filename format is used. Not sure if it is an expected behaviour or not.

### Translations:
```yaml
app:
    allTheThings: "All the things in %localityName%"
```

### Twig:
```yaml
{{ 'app.allTheThings' | trans({
    '%localityName%': locality.getName()
}) }}
```

**Comment by arjenm** [comment]
Well, you did rename the file and thereby started using a completely different 'MessageFormatter'.

But that is something we also realized after I wrote this:
> If you rename it though, no old style transchoice in that file will work, since the new translator simply does not understand it (which is also not documented!).

That statement is true for all translations with variables, including transchoice.

They need to change like so:
> ```yaml
> app:
>     allTheThings: "All the things in {localityName}"
> ```

The {} marks a variable. Plural, select and others are actions upon that variable and are thus included within the {}.

> ### Twig:
> 
> ```yaml
> {{ 'app.allTheThings' | trans({
>     'localityName': locality.getName()
> }) }}
> ```

I'm not sure if {%localityName%} would be valid translation syntax, the intl-system is a bit picky with its variable naming. If that does work, the twig file doesn't need to change (but its a bit ugly to use those old style str_replace-variables syntaxes).
